File: AndroidPromotionSDK/src/main/java/com/ttcn/prm/entry/api/PromotionApiResult.kt
package com.ttcn.prm.entry.api

import com.ttcn.promotionsdk.domain.exception.PromotionErrorCodes

/**
 * Kết quả của [PromotionSDKApi]. Đối ứng 1-1 với `PromotionApiResult.swift` bên iOS, nơi nó là
 * `typealias PromotionApiResult<T> = Result<T, PromotionSDKError>` — Swift đã có `Result` sẵn.
 *
 * API **không ném lỗi nghiệp vụ**: mọi thất bại về [Failure]. Chỉ `CancellationException` thoát ra,
 * để structured concurrency của host còn hoạt động.
 */
sealed interface PromotionApiResult<out T> {

    data class Success<out T>(val data: T) : PromotionApiResult<T>

    data class Failure(val error: PromotionSDKError) : PromotionApiResult<Nothing>
}

/**
 * Lỗi trả về từ SDK. Host `when` trên type này để xử lý từng loại.
 *
 * Bên iOS là `enum PromotionSDKError: Error, LocalizedError` với cùng 8 nhánh; [message] ở đây tương
 * ứng với `errorDescription` bên đó. Sửa một bên thì sửa cả hai.
 *
 * ⚠️ **Thêm nhánh là breaking change.** Host `when` trên type này; Kotlin đòi `when` trên `sealed`
 * phải liệt kê đủ, nên mỗi nhánh mới làm hỏng biên dịch của host (bên iOS enum không `@frozen` cũng
 * vậy). Chốt xong danh sách trước go-live; sau đó thêm nhánh = major bump.
 */
sealed class PromotionSDKError : Exception() {

    abstract override val message: String

    /** Lỗi từ server: [code] là HTTP status nếu có, [message] là mô tả của server. */
    data class NetworkFailure(
        val code: Int?,
        override val message: String,
    ) : PromotionSDKError()

    /** Token hết hạn — host refresh token rồi gọi lại. */
    data object SessionExpired : PromotionSDKError() {
        override val message: String = "Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại."
    }

    data object Timeout : PromotionSDKError() {
        override val message: String = "Yêu cầu bị timeout, vui lòng thử lại."
    }

    /** Server trả `data: null` ở nơi bắt buộc phải có dữ liệu (chi tiết / validate / redemption). */
    data object ParseFailed : PromotionSDKError() {
        override val message: String = "Có lỗi xảy ra với dữ liệu trả về."
    }

    /** Tính năng đang TẮT qua feature flag. Tương ứng mã nghiệp vụ `PRM_MOB_021`. */
    data object FeatureDisabled : PromotionSDKError() {
        override val message: String = "Tính năng ưu đãi hiện đang tạm thời không khả dụng. Vui lòng thử lại sau."
    }

    /**
     * **Lỗi nghiệp vụ của server**, không phải lỗi mạng: [code] là mã server trả (vd
     * `VOUCHER_EXPIRED`), [message] là câu server soạn cho người dùng — `null` khi server không kèm.
     *
     * Trước đây nhánh này bị nhét vào [NetworkFailure] với `message = errorCode`: host nhận một "lỗi
     * mạng" mà thật ra là rule nghiệp vụ, và **mã lỗi thô nằm đúng chỗ đáng lẽ là câu hiển thị cho
     * người dùng** — hiện thẳng lên UI là ra chữ `VOUCHER_EXPIRED`.
     */
    data class BusinessRule(
        val code: String,
        val serverMessage: String?,
    ) : PromotionSDKError() {
        override val message: String get() = serverMessage ?: "Đã có lỗi xảy ra."
    }

    /**
     * Gọi API khi chưa [PromotionSDK.initialize]. Trước đây bên iOS ca này là `preconditionFailure`
     * — SDK làm **crash app của host** vì lỗi thứ tự khởi tạo của host.
     */
    data object NotInitialized : PromotionSDKError() {
        override val message: String = "PromotionSDK chưa được khởi tạo."
    }

    data class Unknown(val error: Throwable) : PromotionSDKError() {
        override val message: String get() = error.message ?: "Đã có lỗi xảy ra."
    }

    /** HTTP status nếu có — chỉ với [NetworkFailure]. */
    val serverCode: Int? get() = (this as? NetworkFailure)?.code

    companion object {
        /**
         * Mã lỗi **thô** (chuỗi) → kiểu lỗi công khai, để host bắt **tường minh** thay vì so chuỗi.
         *
         * Cần hàm này vì SDK có hai bề mặt lỗi không giống nhau:
         * - Headless [PromotionSDKApi] đã trả sẵn [PromotionSDKError] — bắt bằng `is` là xong.
         * - Callback của widget ([com.ttcn.prm.ui.feature.endowview.PRMEndowView.onError],
         *   [PromotionSDK.confirmRedemption]) chỉ trả `String`, mà hằng số mã lỗi nằm trong
         *   `promotionLogic` — module khai `implementation` nên **không có** trên compile classpath
         *   của host. Host không tham chiếu được hằng số, đành hardcode `"PRM_MOB_021"`.
         *
         * Với hàm này host viết được:
         * ```kotlin
         * endowView.onError = { code ->
         *     when (PromotionSDKError.from(code)) {
         *         is PromotionSDKError.FeatureDisabled -> stopCheckout()  // SDK đã tự hiện popup
         *         else -> showMyOwnError()
         *     }
         * }
         * ```
         */
        /**
         * [serverMessage] / [httpStatus]: phần thông tin chỉ có ở bề mặt headless
         * ([PromotionSDKApi] cầm cả `PromotionResult.Failure`). Callback của widget chỉ có mã lỗi
         * nên gọi hàm này với một tham số.
         *
         * **Đây là đường map DUY NHẤT.** Trước đây [PromotionSDKApi] có bản map thứ hai của riêng
         * nó: cùng một lỗi ra hai kết quả khác nhau tuỳ host đi vào đường nào — bản này trả
         * `message = ""`, bản kia trả `failure.message` và cả `httpStatus`.
         */
        @JvmStatic
        @JvmOverloads
        fun from(
            errorCode: String,
            serverMessage: String? = null,
            httpStatus: Int? = null,
        ): PromotionSDKError = when (errorCode) {
            PromotionErrorCodes.FEATURE_DISABLED -> FeatureDisabled
            PromotionErrorCodes.TIMEOUT -> Timeout
            PromotionErrorCodes.NO_RESULT -> ParseFailed
            // Lùi về câu tiếng Việt của SDK khi server không nói gì: đây là case HAY XẢY RA NHẤT, và
            // trước đây nó là case DUY NHẤT trả `message = ""` — host hiện `error.message` thì ra
            // popup trắng không chữ. Chuỗi trùng `R.string.prm_error_network` để hai bề mặt (headless
            // và UI của SDK) không nói hai câu khác nhau cho cùng một sự cố.
            PromotionErrorCodes.NETWORK_ERROR -> NetworkFailure(
                code = httpStatus,
                message = serverMessage?.takeIf { it.isNotBlank() } ?: NETWORK_ERROR_MESSAGE,
            )
            // Mã nghiệp vụ của server (vd VOUCHER_EXPIRED) — KHÔNG phải lỗi mạng.
            else -> BusinessRule(code = errorCode, serverMessage = serverMessage?.takeIf { it.isNotBlank() })
        }

        /** Trùng `R.string.prm_error_network`; để ở đây vì tầng entry không có `Context`. */
        private const val NETWORK_ERROR_MESSAGE =
            "Không có kết nối mạng. Vui lòng kiểm tra rồi thử lại"
    }
}

File: iosPromotionSDK/Entry/API/PromotionApiResult.swift
//
//  PromotionApiResult.swift
//  PromotionSDK
//
//  Đối ứng 1-1 với `PromotionApiResult.kt` bên Android. Sửa một bên thì sửa cả hai.
//

import Foundation
@_implementationOnly import PRMKotlinBridge

/// Kết quả của `PromotionSDKApi`. Bên Android là `sealed interface PromotionApiResult<T>` với hai
/// nhánh `Success` / `Failure`; Swift đã có `Result` nên chỉ cần đặt tên chung để hai tài liệu khớp nhau.
///
/// API **không ném lỗi nghiệp vụ**: mọi thất bại về `.failure`.
public typealias PromotionApiResult<T> = Result<T, PromotionSDKError>

/// Lỗi trả về từ SDK. Đối tác switch trên enum này để xử lý từng loại.
///
/// Bên Android là `sealed class PromotionSDKError : Exception()`, và `errorDescription` ở đây tương
/// ứng với `message` của `Throwable` bên đó.
///
/// ⚠️ **Thêm case là breaking change.** Enum này public, có associated value, và Swift không cho
/// `@frozen` ở module thường — nên mọi `switch` đã liệt kê đủ case bên host sẽ vỡ khi SDK thêm một
/// case. Chốt xong danh sách trước go-live; sau đó thêm case = major bump.
public enum PromotionSDKError: Error, LocalizedError {
    /// Lỗi từ server: `code` là HTTP status nếu có, `message` là mô tả của server.
    case networkFailure(code: Int?, message: String)
    /// Token hết hạn — host refresh token rồi gọi lại.
    case sessionExpired
    case timeout
    /// Server trả `data: null` ở nơi bắt buộc phải có dữ liệu (chi tiết / validate / redemption).
    case parseFailed
    /// Tính năng đang TẮT qua feature flag. Tương ứng mã nghiệp vụ `PRM_MOB_021`.
    case featureDisabled
    /// **Lỗi nghiệp vụ của server**, không phải lỗi mạng: `code` là mã server trả (vd
    /// `VOUCHER_EXPIRED`), `message` là câu server soạn cho người dùng — `nil` khi server không kèm.
    ///
    /// Trước đây nhánh này bị nhét vào `.networkFailure` với `message = errorCode`: host nhận một
    /// "lỗi mạng" mà thật ra là rule nghiệp vụ, và **mã lỗi thô nằm đúng chỗ đáng lẽ là câu hiển thị
    /// cho người dùng** — hiện thẳng lên UI là ra chữ `VOUCHER_EXPIRED`.
    case businessRule(code: String, message: String?)
    /// Gọi API khi chưa `PromotionSDK.initialize`. Trước đây ca này là `preconditionFailure` —
    /// SDK làm **crash app của host** vì lỗi thứ tự khởi tạo của host.
    case notInitialized
    case unknown(Error)

    public var errorDescription: String? {
        switch self {
        case .networkFailure(_, let message): return message
        case .sessionExpired:                 return "Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại."
        case .timeout:                        return "Yêu cầu bị timeout, vui lòng thử lại."
        case .parseFailed:                    return "Có lỗi xảy ra với dữ liệu trả về."
        case .featureDisabled:                return "Tính năng ưu đãi hiện đang tạm thời không khả dụng. Vui lòng thử lại sau."
        case .businessRule(_, let message):   return message ?? "Đã có lỗi xảy ra."
        case .notInitialized:                 return "PromotionSDK chưa được khởi tạo."
        case .unknown(let error):             return error.localizedDescription
        }
    }

    /// HTTP status nếu có — chỉ với `.networkFailure`.
    public var serverCode: Int? {
        if case .networkFailure(let code, _) = self { return code }
        return nil
    }
}

// MARK: - Mã lỗi thô → kiểu lỗi công khai

public extension PromotionSDKError {

    /// Mã lỗi **thô** (chuỗi) → kiểu lỗi công khai, để host bắt **tường minh** thay vì so chuỗi.
    ///
    /// Cần hàm này vì SDK có hai bề mặt lỗi không giống nhau:
    /// - `PromotionSDKApi` (headless) đã trả sẵn `PromotionSDKError` — `switch` là xong.
    /// - Callback của widget (`PromotionSDK.confirmRedemption(onError:)`) chỉ trả `String`, mà hằng
    ///   số mã lỗi nằm trong lõi Kotlin, không phơi ra cho host. Host đành hardcode `"PRM_MOB_021"`.
    ///
    /// ```swift
    /// PromotionSDK.confirmRedemption(onSuccess: { … }, onError: { code in
    ///     switch PromotionSDKError.from(code) {
    ///     case .featureDisabled: stopCheckout()   // SDK đã tự hiện popup, host chỉ cần dừng luồng
    ///     default:               showMyOwnError()
    ///     }
    /// })
    /// ```
    ///
    /// Đối ứng `PromotionSDKError.from(errorCode)` bên Android — sửa một bên thì sửa cả hai.
    /// - Parameters:
    ///   - serverMessage: câu của server, chỉ có ở bề mặt headless. Callback của widget chỉ có mã.
    ///   - httpStatus: HTTP status, cùng lý do.
    static func from(_ errorCode: String,
                     serverMessage: String? = nil,
                     httpStatus: Int? = nil) -> PromotionSDKError {
        switch errorCode {
        case PromotionErrorCodes.shared.FEATURE_DISABLED: return .featureDisabled
        case PromotionErrorCodes.shared.TIMEOUT:          return .timeout
        case PromotionErrorCodes.shared.NO_RESULT:        return .parseFailed
        // Lùi về câu tiếng Việt của SDK khi server không nói gì: đây là case HAY XẢY RA NHẤT, và
        // trước đây là case DUY NHẤT trả `message: ""` — host hiện `error.localizedDescription` thì
        // ra popup trắng không chữ. Chuỗi trùng `PromotionUIStrings` để hai bề mặt (headless và UI
        // của SDK) không nói hai câu khác nhau cho cùng một sự cố.
        case PromotionErrorCodes.shared.NETWORK_ERROR:
            let message = serverMessage?.trimmingCharacters(in: .whitespaces)
            return .networkFailure(code: httpStatus,
                                   message: (message?.isEmpty == false ? message! : networkErrorMessage))
        // Mã nghiệp vụ của server (vd VOUCHER_EXPIRED) — KHÔNG phải lỗi mạng.
        default:
            let message = serverMessage?.trimmingCharacters(in: .whitespaces)
            return .businessRule(code: errorCode, message: message?.isEmpty == false ? message : nil)
        }
    }

    /// Trùng `R.string.prm_error_network` bên Android — sửa một bên thì sửa cả hai.
    private static var networkErrorMessage: String {
        "Không có kết nối mạng. Vui lòng kiểm tra rồi thử lại"
    }
}
